support extend inside a message body, rough cut

mmmries opened the pull request:
This is a rough cut at resolving #42. I've added a test where the message gets extended inside its own definition and gotten the test to pass.
My biggest concern at this point is that I'm manually qualifying the message name (making it a root message name) because I wasn't sure of the right way to qualify it without applying the message name that it is nested inside of. 
Can anyone point me towards a better way of handling the message name qualification?

mmmries commented on src/gpb_parse.yrl:
```diff
@@ -410,6 +412,9 @@ flatten_fields(FieldsOrDefs, FullName) ->
                        (#gpb_oneof{}=O, {Fs,Ds}) ->
                             {[O | Fs], Ds};
+                       ({{extend, Msg},Exts}, {Fs,Ds}) ->
+                            QExt = {{extend, ['.',Msg]}, Exts},
```
This is the manual qualification that I'm not comfortable with. If this `extend` was happening inside a message with a `FullName = ['.','some','.','package','.','m1']` I think it should have a qualified name of `['.', 'some', '.', 'package', '.', Msg]`

tomas-abrahamsson replied:
Excellent! I think you're right in concern about qualification. With a small change of your clause in `flatten_fields` into this...
```erlang
                       ({{extend, _Msg},_Exts}=Def, {Fs,Ds}) ->
                            QDefs = flatten_qualify_defnames(
                                      [Def], drop_last_level(FullName)),
                            {Fs, QDefs++Ds};
```
...then the following extra test will also pass:
```erlang
parses_nested_extending_msgs_in_package_test() ->
    {ok,Defs} = parse_lines(["package p1.p2;",
                             "message m1 {",
                             "  required uint32 f1=1 [default=17];",
                             "  extensions 200 to 299;",
                             "  extend m1 {",
                             "    optional uint32 f2=2;",
                             "  }",
                             "}"]),
    [{package,'p1.p2'},
     {{extensions,'p1.p2.m1'},[{200,299}]},
     {{msg,'p1.p2.m1'},
      [#?gpb_field{name=f1, fnum=1, rnum=2, opts=[{default,17}],
                   occurrence=required},
       #?gpb_field{name=f2, fnum=2, rnum=3, opts=[],
                   occurrence=optional}]}] =
        do_process_sort_defs(Defs, [use_packages]).
```
I can amend your commit with this, if you want(?) or if you want to update the commit and force-push yourself? Whichever way is fine for me

mmmries commented:
I'll update the PR quickly with your suggested fix. Thanks for checking this out so quickly.
Really looking forward to getting some erlang and elixir in production :heart: :+1:

mmmries commented:
I added your test case and suggested fix and squashed the two commits to avoid cluttering the git history.

```
L422-MRies:gpb mmmries$ make test
erlc -o src src/gpb_parse.yrl
erlc -Wall +debug_info -Iinclude -pa ebin -o ebin src/gpb_parse.erl
Proto-compiling the description definition...
erlc -Wall +debug_info -Iinclude -pa ebin -o ebin descr_src/gpb_compile_descr.erl
erlc -Wall +debug_info -Iinclude -pa ebin -o ebin descr_src/gpb_descriptor.erl
erlc -Wall +debug_info -Iinclude  -pa ebin -o test test/gpb_parse_tests.erl
Testing...
  All 296 tests passed.
```

tomas-abrahamsson commented:
Brilliant! The commit message metions "rough cut", I guess it is no longer that rough :) Ok if I go ahead and just remove the rough cut phrase?

mmmries commented:
sure, you can edit the commit message to make it fit the project styles and history. Full editor privileges :)

tomas-abrahamsson commented:
Thanks! I've just pushed (and also to hex.pm, needed to update some libs first)
